### ISSUE TITLE
Feature: Strided convolutions

### ISSUE THREAD

**Issue by cpdurham** [AccelerateHS/accelerate#387]
Something like this would be amazing (I can't figure out how to do it efficiently in accelerate proper)
``` haskell
stridedConvolution
  :: Exp sh --stride
  -> Exp sh --padding?
  -> (Exp a -> Exp b -> Exp c) --multiply
  -> (Exp c -> Exp c -> Exp d) --add
  -> Array sh b --kernel
  -> Array sh a --source
  -> Array sh d
```

**Comment by tmcdonell** [comment]
Kind of like a `stencil`, but only applied to a subset of the elements (like in a checkerboard pattern say)?

**Comment by cpdurham** [comment]
Exactly, if a kernel tile was say 5x5, then a stride of 5x5 would cover each block in a checkerboard with no overlap. Where a stride of 10x10 would cover every black square on a checker board with every other row turned all white.

**Comment by cpdurham** [comment]
This is great, sorry just found it https://github.com/vdumoulin/conv_arithmetic

**Comment by tmcdonell** [comment]
Interesting! So in this description it looks like Accelerate's `stencil` corresponds to "half padding, no strides". I hadn't thought about all the other kinds before.

**Comment by cpdurham** [comment]
Yeah, deep learning brought about a lot of the interest recently. Conv transpose or deconv is responsible for most of the neural nets that generate data or unsupervised learning. Dilations are the new hot thing right now for networks to gain a larger receptive field per layer. One of the best neural net architectures (ResNet) uses 7x7 convolutions with stride 3x3. Also max pooling is used a lot ('multiply' is constant in source image, 'add' is max).
